app/models/measurement/gr_value_loader.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+class Measurement
+  class GrValueLoader
+    include ActiveModel::Model
+    include ActiveRecord::AttributeAssignment
+
+    attr_accessor :measurement, :levels, :ministry_id, :assignment_id,
+                  :mcc, :period, :source, :historical
+
+    def load_gr_value
+      return load_historic_gr_values if @historical
+      @levels.each do |level|
+        measurement_level_id = @measurement.send("#{level}_id")
+        filter_params = gr_request_params(level)
+        gr_resp = GlobalRegistryClient.client(:measurement_type).find(measurement_level_id, filter_params)
+        value = gr_resp['measurement_type']['measurements'].map { |m| m['value'].to_f }.sum
+        @measurement.send("#{level}=", value)
+      end
+    end
+
+    private
+
+    def load_historic_gr_values
+      return unless can_historic
+      gr_resp = GlobalRegistry::MeasurementType
+                .find(@measurement.total_id, gr_request_params(:total))['measurement_type']['measurements']
+      @measurement.total = build_total_hash(gr_resp)
+    end
+
+    def build_total_hash(gr_resp)
+      total_hash = {}
+      i_period = period_from
+      loop do
```

Comment:
It took me a little bit of thinking to mentally parse out what this loop is doing. Now I think I understand it: it's looping over all of the months from `period_from` to `@period` and for each month in that range, it's adding that period month to `total_hash` as an entry.

I notice that the ending point is `i_period == @period` - is there any case when that might not occur and this might be an infinite loop?

And I definitely don't want you to feel like you're burning time on this so if you think it's too much that's OK, but what would you think about making the iteration index of the loop a bit clearer, something like figuring out the difference in months first and then looping over that with a month index?

That would also make it have a definite and obvious endpoint even if an infinite loop would never happen in practice.